snuba/replacers/replacer_processor.py

```diff
@@ -16,14 +17,36 @@ class ReplacementMessage(NamedTuple):
     data: Mapping[str, Any]
 
 
-@dataclass(frozen=True)
-class Replacement:
-    # XXX: For the group_exclude message we need to be able to run a
-    # replacement without running any query.
-    count_query_template: Optional[str]
-    insert_query_template: Optional[str]
-    query_args: Mapping[str, Any]
-    query_time_flags: Any
+class Replacement(ABC, Generic[C]):
+    @abstractclassmethod
+    def parse_message(
+        cls, message: Mapping[str, Any], context: C
+    ) -> Optional["Replacement[C]"]:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def get_project_id(self) -> int:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def get_excluded_groups(self) -> Sequence[int]:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def get_needs_final(self) -> bool:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def get_insert_query_template(self) -> Optional[str]:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def get_count_query_template(self) -> Optional[str]:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def get_query_args(self) -> Mapping[str, Any]:
+        raise NotImplementedError()
```

Comment:
Is this the part you were referring to as 

> ...though the interface may have to change in a bunch of ways to
actually solve the problems that Filippo complains about wrt poor typing
of queries. The important bit is that replacements now encapsulate their
own logic.

I think we could get rid of this template entirely and let the instances of this class produce a query given some parameters provided by the replacer processor. But I would be ok in doing that in a second step and starting by reorganizing the code.

Reply:
could you elaborate what you want to do here? specifically, how do you want to communicate data between ReplacerWorker that needs those attrs and ReplacerProcessor implementors

Reply:
This is what we do today:
https://github.com/getsentry/snuba/blob/master/snuba/replacer.py#L53-L62

We return a template + a dictionary of arguments to replace in the template. The arguments are `replacement.query_args`. Some arguments are populated by replacer.py like `table_name`.

This creates a lot of coupling between the two classes. replacer.py relies on the template structure and replacer_processor relies on replacer.py to fill in a missing parameter. All this is not typed.

What I think we could do is to make the `Replacement` class provide methods that just return the SQL query:

```
class Replacer(ABC):
   def get_count_query(self, table_name: str):
       ....

   def get_insert_query(self, table_name: str):
       ....
```

So only the Replacement class will deal with the template and will never expose it. This makes this logic easy to test and reduces coupling. 

Though it makes sense to me to do such change in a separate PR to reduce the scope